Add Certificate Transparency support

aeijdenberg opened the pull request:
First of a series of commits that add Certificate Transparency support to OpenSSL.  See: http://www.certificate-transparency.org/

Original authors:
Rob Stradling <rob@comodo.com>
Dr. Stephen Henson <steve@openssl.org>

ghedo commented:
Aren't the `SCT_*()` functions supposed to be public? Then maybe a new `include/openssl/ct.h` header is a better place where to declare them.

Also, it seems that there are several unrelated changes included (e.g. ssl/Makefile, test/Makefile, util/libeay.num ...). Maybe they should go into separate commits.

IMO the commit message could probably be improved as well....

aeijdenberg commented:
Hi Alessandro,

I just added some additional context on the list.  For now I deliberately did not make the SCT_* functions public.  We intend to eventually, however there are some reasons to wait a little while, e.g. for RFC6962-bis process to complete.  The intent of this patch is to form a lower level of base functionality that a higher level API will make use of (link sent to more info on this list).

With regards to the other Makefile, libeay.num changes - how should one best handle the modifications that "make update" seems to make (such as those)?

Cheers, Adam

richsalz commented:
With regards to the other Makefile, libeay.num changes - how should one best handle the modifications that "make update" seems to make (such as those)?

Yeah, it's a pain.  Do "make update" as a separate commit, I guess.

aeijdenberg commented:
Rebased to put the make update churn in a separate commit.

aeijdenberg commented:
Emilia, I made updates based on your comments.  I (perhaps foolishly) rebased and squashed them so as to make them easier to read, and looks like GitHub has eaten your comments as a consequence.  Sorry about that.

Basically I think I fixed everything you were looking for, and removed the key duplication method for now - we can revisit once we commit code that needs to actually call it.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,301 @@
+
+int SCT_set_version(SCT *sct, sct_version_t version)
+{
+    if (sct == NULL) {
```
The team consensus was towards not checking NULL inputs, particularly for the |foo| in a FOO_xxx API. This is (mostly) consistent with existing OpenSSL style, and also akin to standard library.

So you can remove these checks.

ekasper replied:
(This also means that some, though not all of them, can become void.)

aeijdenberg replied:
I've removed instances of if (sct == NULL) from SCT_xxx functions.
I kept the NULL checks for pointers that are populated (such as the SCT_get_xxx calls).  Let me know if that is what you expected or not.
I don't understand your comment about "becoming void" though, can you please elaborate?

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,301 @@
+    else if (nid == NID_ecdsa_with_SHA256)
+        sct->sig_alg = TLSEXT_signature_ecdsa;
+    else
+        return 0;
```
Set a CTerr.

aeijdenberg replied:
Fixed.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,301 @@
+    return 1;
+}
+
+int SCT_get_version(const SCT *sct, sct_version_t *version)
```
As above, remove the NULL checks. This means you can simplify
`sct_version_t SCT_get_version(const SCT *sct)`
etc

aeijdenberg replied:
Fixed.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,301 @@
+    return 1;
+}
+
+int SCT_get0_logid(const SCT *sct, unsigned char **logid, size_t *logidlen)
```
These, too, can just make it part of the API contract that |logid| and |logidlen| must not be NULL, and become void.

aeijdenberg replied:
Updating to return the size instead of an error code.  Let me know what you think.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,301 @@
+        if (sct->sig_alg == TLSEXT_signature_ecdsa)
+            return NID_ecdsa_with_SHA256;
+    }
+    return NID_undef;
```
The current v3_scts.c code will set hash_alg and sig_alg without any checks, so invalid values can enter the system. I imagine that, for debugging and pretty-printing, you'll want to try to recognize those, rather than return NID_undef.

But that should maybe happen in a different API call. Let's see how this API gets used; fine to leave as is for now.

aeijdenberg replied:
Ack.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,301 @@
+    }
+    if (sct->version != SCT_V1) {
+        CTerr(CT_F_SCT_GET_SIGNATURE_NID, CT_R_SCT_UNSUPPORTED_VERSION);
+        return -1;
```
I don't particularly like returning -1 here, even if it's sort of an undocumented contract that NIDs are positive. Maybe just return NID_undef.

aeijdenberg replied:
Fixed.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,301 @@
+    if (sct->version == UNSET_VERSION)
+        return 0;
+    if (sct->version != SCT_V1) {
+        /* Just need cached encoding */
```
I'd prefer sct_check_format to do what it does, i.e., check the format. If something that calls this can operate only with the cached encoding, then this should be a separate check.

Maybe remove from this MR and add back when it's actually called, so we see how it's used.

aeijdenberg replied:
Removed as requested.

ekasper commented on crypto/ct/ct_locl.h:
```diff
@@ -0,0 +1,266 @@
+     */
+    unsigned char *logid;
+    size_t logidlen;
+    uint64_t timestamp;
```
This doesn't let you distinguish an unset timestamp from a 0 timestamp. I kinda think it's fine - no legit SCT should have a 0 timestamp - but maybe just throw a comment somewhere.

aeijdenberg replied:
Added comment.

ekasper commented on crypto/ct/ct_locl.h:
```diff
@@ -0,0 +1,266 @@
+
+typedef struct {
+    sct_version_t version;
+    /* If version is not 0 this contains the encoded SCT */
```
nit: "If version is not SCT_V1"

aeijdenberg replied:
Fixed.

ekasper commented on crypto/ct/ct_locl.h:
```diff
@@ -0,0 +1,266 @@
+
+/*
+ * Allocate new SCT.
+ * Caller is response to call SCT_free when done.
```
response to call -> responsible for calling

aeijdenberg replied:
Fixed.

ekasper commented on crypto/ct/ct_locl.h:
```diff
@@ -0,0 +1,266 @@
+
+/*
+ * Set the version of an SCT.
+ * Returns 1 on success.
```
1 on success -> 1 on success, 0 if the version is unrecognized.

aeijdenberg replied:
Fixed.

ekasper commented on crypto/ct/ct_locl.h:
```diff
@@ -0,0 +1,266 @@
+# define CT_F_SCT_SET_VERSION                             111
+
+/* Reason codes. */
+# define CT_R_ILLEGAL_CURVE                               100
```
Some of those codes aren't used now, would be nice to clean this up and bring them back as they're needed.

aeijdenberg replied:
Fixed.

aeijdenberg commented:
Rebased against current master.

aeijdenberg commented:
Another rebase to pick up the fixes to master to allow a clean build.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,222 @@
+
+void sct_free_internal(SCT *sct)
+{
+    if (sct->logid)
```
Missed this before: OPENSSL_free is NULL-safe, so you can remove these if's.

aeijdenberg replied:
Fixed.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,222 @@
+
+void SCT_free(SCT *sct)
+{
+    if (sct) {
```
Here, too, no need for the if ().

aeijdenberg replied:
True for the second statement, but not the first.  If we don't check NULL, then we will try to dereference NULL in sct_free_internal.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,222 @@
+    return 1;
+}
+
+int SCT_set0_logid(SCT *sct, void *logid, size_t logidlen)
```
nit: log_id_len would be a bit easier on the eye

aeijdenberg replied:
Fixed across the board.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,222 @@
+        CTerr(CT_F_SCT_SET0_LOGID, CT_R_INVALID_LOGID_LENGTH);
+        return 0;
+    }
+    if (sct->logid)
```
No if.

aeijdenberg replied:
Fixed.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,222 @@
+    return 1;
+}
+
+int SCT_set_timestamp(SCT *sct, uint64_t timestamp)
```
This can be void, since it always succeeds.

Unless you want to leave the door open for future SCT versions that don't have a timestamp field? But given that an SCT is a Signed Certificate *Timestamp*, I find that... unlikely?

ekasper replied:
Ditto about extensions and signature.

aeijdenberg replied:
Fixed.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,222 @@
+    }
+}
+
+int SCT_set0_extensions(SCT *sct, void *ext, size_t extlen)
```
Why void? unsigned char *ext I think.

ekasper replied:
(everywhere)

aeijdenberg replied:
Fixed, I had changed to unsigned char perhaps after misunderstanding a previous comment.

ekasper commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,222 @@
+                return NID_ecdsa_with_SHA256;
+            case TLSEXT_signature_rsa:
+                return NID_sha256WithRSAEncryption;
+            }
```
I would add a
default:
    return NID_undef
to make it more explicit.

aeijdenberg replied:
Fixed.

aeijdenberg commented:
Emilia, thanks for the comments.  Have pushed another commit that I believe addresses the outstanding issues.

richsalz commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,213 @@
+    return sct;
+}
+
+void sct_free_internal(SCT *sct)
```
I'd prefer this to be inlined into SCT_free, it keeps the parts together.

aeijdenberg replied:
Fixed.

richsalz commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,213 @@
+    return 1;
+}
+
+void SCT_set_timestamp(SCT *sct, uint64_t timestamp)
```
I know 64bit is what's in the RFC's, but should we have an openssl version that takes ASN1_time?  Perhaps not, but I'm just asking.

richsalz commented on crypto/ct/ct_lib.c:
```diff
@@ -0,0 +1,213 @@
+
+int SCT_get_signature_nid(const SCT *sct)
+{
+    /* RFC6962 only permits two signature algorithms */
```
not sure that comment is needed.  i first thought it meant "two at once"

aeijdenberg replied:
Fixed.

aeijdenberg commented:
Closing PR as ekasper@ has committed the changes.  Will open a new PR for the next batch.
